### PR TITLE
Update FFI specific code for compilation in JDK21

### PR DESCRIPTION
**Pull request by ChengJin01**
The changes update FFI specific code by disabling/removing
part of our code to pass the compilation in JDK21.

Note:
The modifications in code are based on the new APIs in JDK21 & JEP442
as mentioned at https://github.com/eclipse-openj9/openj9/issues/16951.

Signed-off-by: ChengJin01 <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
Note:
1) All `VaList` related code are removed as required in JEP442 at https://github.com/eclipse-openj9/openj9/issues/16951
2)  The PR must be merged along with the disabled code in OpenJ9 at https://github.com/eclipse-openj9/openj9/pull/17352.

File: src/java.base/share/classes/jdk/internal/foreign/SystemLookup.java [hunk 1]
```diff
@@ -64,11 +65,13 @@ private SystemLookup() { }
...
         try {
-            return switch (CABI.current()) {
-                case SYS_V, LINUX_AARCH_64, MAC_OS_AARCH_64, LINUX_RISCV_64, SYS_V_PPC_64LE, SYS_V_S390X -> libLookup(libs -> libs.load(jdkLibraryPath("syslookup")));
-                case AIX_PPC_64 -> makeAixLookup();
-                case WIN_64, WIN_AARCH_64 -> makeWindowsLookup(); // out of line to workaround javac crash
-            };
+            if (Utils.IS_AIX) {
+                return makeAixLookup();
+            } else if (Utils.IS_WINDOWS) {
+                return makeWindowsLookup();
```
Comment (keithc-ca) [inline src/java.base/share/classes/jdk/internal/foreign/SystemLookup.java:71]: Please add the AIX case after the Windows case to minimize the change relative to upstream:
```java
            if (Utils.IS_WINDOWS) {
                return makeWindowsLookup();
            } else if (Utils.IS_AIX) {
                return makeAixLookup();
```

File: src/java.base/share/classes/jdk/internal/foreign/SystemLookup.java [hunk 2]
```diff
@@ -100,12 +103,11 @@ private static SymbolLookup makeAixLookup() {
                  */
                 SymbolLookup funcsLibLookup =
                         libLookup(libs -> libs.load(jdkLibraryPath("syslookup")));
-                MemorySegment funcs = MemorySegment.ofAddress(funcsLibLookup.find("funcs").orElseThrow().address(),
-                    ADDRESS.byteSize() * AixFuncSymbols.values().length, SegmentScope.global());
+                MemorySegment funcs = MemorySegment.ofAddress(funcsLibLookup.find("funcs").orElseThrow().address());
```
Comment (keithc-ca) [inline src/java.base/share/classes/jdk/internal/foreign/SystemLookup.java:106]: Suggest this should follow the pattern used for Windows:
```java
                MemorySegment funcs = funcsLibLookup.find("funcs").orElseThrow()
                        .reinterpret(AixFuncSymbols.LAYOUT.byteSize());
```
This, of course, requires a definition of `AixFuncSymbols.LAYOUT`:
```java
        static final SequenceLayout LAYOUT = MemoryLayout.sequenceLayout(
                values().length, ADDRESS);
```

File: src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java [hunk 3]
```diff
@@ -37,6 +37,7 @@
 import jdk.internal.foreign.abi.aarch64.windows.WindowsAArch64Linker;
 import jdk.internal.foreign.abi.ppc64.aix.AixPPC64Linker;
 import jdk.internal.foreign.abi.ppc64.sysv.SysVPPC64leLinker;
+import jdk.internal.foreign.abi.fallback.FallbackLinker;
```
Comment (keithc-ca) [inline src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java:40]: Please adjust imports so additions for OpenJ9 keep things ordered.

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java [hunk 4]
```diff
@@ -32,6 +32,7 @@
 
 package jdk.internal.foreign.abi.aarch64;
 
+import java.lang.foreign.AddressLayout;
```
Comment (keithc-ca) [inline src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:35]: Please remove
```java
import sun.security.action.GetPropertyAction;
```
and
```java
    private static final boolean isWinOS = GetPropertyAction.privilegedGetProperty("os.name").startsWith("Windows");
```
replacing `isWinOS` with `Utils.IS_WINDOWS`.

File: src/java.base/share/classes/jdk/internal/foreign/layout/ValueLayouts.java [hunk 5]
```diff
@@ -204,268 +199,240 @@ final V self() {
...
-        public OfDoubleImpl withOrder(ByteOrder order) {
-            Objects.requireNonNull(order);
-            return new OfDoubleImpl(order, bitAlignment(), name());
+        OfDoubleImpl dup(ByteOrder order, long bitAlignment, Optional<String> name) {
+            return new OfDoubleImpl(order, bitAlignment, name);
         }
 
         public static OfDouble of(ByteOrder order) {
-            return new OfDoubleImpl(order);
+            return new OfDoubleImpl(order, isAixOS ? 32 : Double.SIZE, Optional.empty());
```
Comment (keithc-ca) [inline src/java.base/share/classes/jdk/internal/foreign/layout/ValueLayouts.java:324]: Please remove `isAixOS` and use `Utils.IS_AIX` instead.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/jdk/internal/foreign/SystemLookup.java (hunk 1 above)
Comment (ChengJin01) [reply src/java.base/share/classes/jdk/internal/foreign/SystemLookup.java]: Fixed as suggested above.

File: src/java.base/share/classes/jdk/internal/foreign/SystemLookup.java (hunk 2 above)
Comment (ChengJin01) [reply src/java.base/share/classes/jdk/internal/foreign/SystemLookup.java]: Agreed and updated the code as suggested above.

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java (hunk 4 above)
Comment (ChengJin01) [reply src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:35]: Fixed as suggested above.

File: src/java.base/share/classes/jdk/internal/foreign/layout/ValueLayouts.java (hunk 5 above)
Comment (ChengJin01) [reply src/java.base/share/classes/jdk/internal/foreign/layout/ValueLayouts.java]: Fixed as suggested above.

File: src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java (hunk 3 above)
Comment (ChengJin01) [reply src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java]: Fixed as suggested above.

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java (hunk 4 above)
Comment (keithc-ca) [reply src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:35]: The (unused) import is still present.

Comment (ChengJin01) [reply src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:35]: removed `import sun.security.action.GetPropertyAction;`.

**Comment by keithc-ca** [comment]
Jenkins compile all jdknext depends https://github.com/eclipse-openj9/openj9/pull/17352

**Comment by ChengJin01** [comment]
The compilation failure was detected at https://openj9-jenkins.osuosl.org/job/Build_JDKnext_x86-64_linux_Personal/66/consoleText and https://openj9-jenkins.osuosl.org/job/Build_JDKnext_s390x_linux_Personal/88/consoleText due to the missing method `clearInetAddressCache` as follows:
```
[2023-05-29T21:02:15.177Z] /home/jenkins/workspace/Build_JDKnext_x86-64_linux_Personal/build/linux-x86_64-server-release/support/j9jcl/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:624: error: cannot find symbol
[2023-05-29T21:02:15.177Z] 		SharedSecrets.getJavaNetInetAddressAccess().clearInetAddressCache();
[2023-05-29T21:02:15.177Z] 		                                           ^
[2023-05-29T21:02:15.177Z]   symbol:   method clearInetAddressCache() <--------------------
[2023-05-29T21:02:15.177Z]   location: interface JavaNetInetAddressAccess
[2023-05-29T21:02:15.177Z] 1 error
```
which was introduced recently via https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/604 which is not yet added to the `openj9-staging` branch.

@JasonFengJ9, could you help add your changes at https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/604 to your branch at `https://github.com/JasonFengJ9/openj9-openjdk-jdk/tree/mergetmp?  I will rebase with your updated code from there.

**Comment by keithc-ca** [comment]
@ChengJin01 It might be simpler to just cherry-pick f68f90bc852aae0f474ae6eb824c268533c9f339 into your branch.

**Comment by ChengJin01** [comment]
> > It might be simpler to just cherry-pick [f68f90b](https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/f68f90bc852aae0f474ae6eb824c268533c9f339) into your branch.
> 
> Agreed, @ChengJin01 let me know if the initial merging PR rebase is still needed.

It doesn't help as my branch was created on top of https://github.com/JasonFengJ9/openj9-openjdk-jdk/tree/mergetmp in which case the expected changes at https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/f68f90bc852aae0f474ae6eb824c268533c9f339 doesn't show up in the history/annotation list for cherry-pick.

**Comment by ChengJin01** [comment]
I've cherry-picked the changes at https://github.com/ibmruntimes/openj9-openjdk-jdk/commit/f68f90bc852aae0f474ae6eb824c268533c9f339#diff-bcf494d3974af1421572e5bf8d544e4949a90eacb34c4fcbc4e4668a0a89da6b from the `openj9` branch.

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**